### PR TITLE
Cpp14 support

### PR DESCRIPTION
**Pull request by oliverlee**
I've updated the P0009 headers to compile without C++17 extensions.

For the last commit, I've dumped a few common utilities in the extents header. Please feel free to edit as desired.

### CODE REVIEW

File: include/experimental/__p0009_bits/extents.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]: Nothing in this header threw an `std::exception` subclass before.  Furthermore, the only reason this code now throws an exception is to trigger a call to `terminate` in a `noexcept` function.

Would you consider just calling `std::abort` as before?  Alternately, would you consider adding a configurable precondition violation handler (e.g., a macro that takes a message and just calls `std::abort` by default)?

File: include/experimental/__p0009_bits/extents.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/extents.hpp]: Would you consider adding a nonempty message here?

File: include/experimental/__p0009_bits/fwd.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/fwd.hpp]: Would you consider making this an alias of `std::integral_constant`, so as to avoid more instantiations of types?

**Comment by mhoemmen** [comment]
Thanks for the contribution!  There are some nice refactorings in there as well.  @crtrott et al. will likely want to comment.

1. Do mdspan's tests and examples all build with C++14, or do these changes only just make your application build?  We're generally only interested in the former.

2. Would you consider adding C++14 testing to https://github.com/kokkos/mdspan/blob/stable/.github/workflows/cmake.yml as a part of this PR?  Current PR automatic testing will not exercise C++14.

File: include/experimental/__p0009_bits/extents.hpp
Comment (oliverlee) [reply include/experimental/__p0009_bits/extents.hpp]: > Would you consider just calling std::abort as before? Alternately, would you consider adding a configurable precondition violation handler (e.g., a macro that takes a message and just calls std::abort by default)?

Sure that sounds like a better approach. I'm also wondering if `extents.hpp` is the right location for the `terminate_if_invalid_strides` function. Maybe `layout_stride.hpp` is a better place?

**Comment by oliverlee** [comment]
>     1. Do mdspan's tests and examples all build with C++14, or do these changes only just make your application build?  We're generally only interested in the former.

I've only tested with the tests defined in CMake along with C++14 so far. All tests pass before and after this PR - however after this PR now builds without warnings.

>     2. Would you consider adding C++14 testing to https://github.com/kokkos/mdspan/blob/stable/.github/workflows/cmake.yml as a part of this PR?  Current PR automatic testing will not exercise C++14.

Sure.

**Comment by mhoemmen** [comment]
Thanks for making the changes!  Building with C++14 results in the following CMake error message.

```
CMake Error at CMakeLists.txt:200 (MESSAGE):
  Benchmarks are not available in C++14 or C++23 mode.  Turn
  MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.
```

I think you could just change line 200 of CMakeLists.txt ( https://github.com/kokkos/mdspan/blob/0e6a69dfe045acbb623003588a4aff844ea4b276/CMakeLists.txt#L200 ) from what it is currently,

```
MESSAGE(FATAL_ERROR "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
```

to something like

```
set(MDSPAN_ENABLE_BENCHMARKS OFF)
MESSAGE(STATUS "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
```

**Comment by oliverlee** [comment]
> Thanks for making the changes! Building with C++14 results in the following CMake error message.
> 
> ```
> CMake Error at CMakeLists.txt:200 (MESSAGE):
>   Benchmarks are not available in C++14 or C++23 mode.  Turn
>   MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.
> ```
> 
> I think you could just change line 200 of CMakeLists.txt (
> 
> https://github.com/kokkos/mdspan/blob/0e6a69dfe045acbb623003588a4aff844ea4b276/CMakeLists.txt#L200
> ) from what it is currently,
> 
> ```
> MESSAGE(FATAL_ERROR "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
> ```
> 
> to something like
> 
> ```
> set(MDSPAN_ENABLE_BENCHMARKS OFF)
> MESSAGE(STATUS "Benchmarks are not available in C++14 or C++23 mode. Turn MDSPAN_ENABLE_BENCHMARKS OFF or use C++17 or C++20.")
> ```

I disabled benchmarking for C++14 using the same approach for C++23 (i.e. via matrix settings in `cmake.yml`) since I thought it would be simpler to maintain. Let me know if you still prefer a CMake option based approach.

File: include/experimental/__p0009_bits/extents.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/extents.hpp]:
```suggestion
      rank() == extents<OtherIndexType, OtherExtents...>::rank() &&
```

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/macros.hpp]:
```suggestion
#ifndef MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER
#define MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
```

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/macros.hpp]:
```suggestion
#ifndef MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER
  #ifndef NDEBUG
    #define MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER 0
  #else
    #define MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER 1
  #endif
#endif
```

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/macros.hpp]:
```suggestion
template <bool skip = MDSPAN_IMPL_SKIP_PRECONDITION_VIOLATION_HANDLER>
```

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/macros.hpp]:
```suggestion
#define MDSPAN_IMPL_PRECONDITION(...) \
```

File: include/experimental/__p0009_bits/layout_stride.hpp
Comment (crtrott) [inline include/experimental/__p0009_bits/layout_stride.hpp]:
```suggestion
    MDSPAN_IMPL_PRECONDITION(common_integral_compare(stride, other.stride(s))
                             and "invalid strides for layout_{left,right}");
```

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (crtrott) [inline tests/test_alternate_precondition_violation_handler.cpp]:
```suggestion
#define MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line) \
  do { \
    throw std::logic_error{"precondition failure"}; \
  } while (0);
```

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (crtrott) [inline tests/test_alternate_precondition_violation_handler.cpp]:
```suggestion
  ASSERT_THROW(MDSPAN_IMPL_PRECONDITION(false), std::logic_error);
```

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (crtrott) [inline tests/test_alternate_precondition_violation_handler.cpp]:
```suggestion
      MDSPAN_IMPL_PRECONDITION(1 + 1 == 2);
```

File: tests/test_macros.cpp
Comment (crtrott) [inline tests/test_macros.cpp]:
```suggestion
  ASSERT_DEATH(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"), msg);
```

File: tests/test_macros.cpp
Comment (crtrott) [inline tests/test_macros.cpp]:
```suggestion
      MDSPAN_IMPL_PRECONDITION(1 + 1 == 2);
```

**Review by crtrott — CHANGES_REQUESTED** [review]
This looks pretty good. Finally found sufficient time to review it. The new macros are not a bad idea, but I think they should all be IMPL macros.

**Comment by crtrott** [comment]
I will rebase this and merge my suggestions.

**Comment by crtrott** [comment]
@oliverlee FYI: I rebased and force pushed. I'd like to get this in before I work on more of the features for C++26.

**Review by crtrott — APPROVED** [review]

**Comment by oliverlee** [comment]
Not sure why the death tests are failing now - they were fine a few months back.

I'll probably have time to look into it tomorrow.

**Comment by crtrott** [comment]
They passed now. We might have changed Debug to Release build at some point and so the rebase made them fail.

**Comment by oliverlee** [comment]
It should be possible to enable precondition checks in release mode with a preprocessor macro - I imagine this is enabled only for tests and disabled for benchmarks.

It shouldn't be too much trouble to do in a separate PR if you want to merge this PR and use it as a base for c++26 stuff.

**Comment by oliverlee** [comment]
@crtrott It looks like the macro in `test/CMakeLists.txt` was missed in the rename commit.

I've created two commits - please squash or do whatever is normal for this repo.

**Comment by crtrott** [comment]
Thanks! I had overlooked that. That said I am not sure it's the best way to go about this. Now the tests never take the code path where the preconditions are skipped. Maybe instead we should have a mix of debug and non-debug test configs.

**Comment by crtrott** [comment]
Also I am waiting for a second review to finish (should happen latest by Monday I pinged appropriate people in my team).

**Comment by oliverlee** [comment]
> That said I am not sure it's the best way to go about this. 

Same - whatever makes the most sense to you in terms of maintenance.

I've seen precondition tests defined in a separate test file. That could remove the need to conditionally define test cases with `ASSERT_DEATH` and simplify running tests with debug and non-debug configs. Would that make sense here?

**Comment by crtrott** [comment]
It could make sense to separate them out. We had some platforms where the death tests never fully worked ...

**Comment by oliverlee** [comment]
> It could make sense to separate them out. We had some platforms where the death tests never fully worked ...

There are only a few `DEATH_TESTS` now.

It makes sense to do it earlier rather than later if this is approach you want to pursue.

**Comment by crtrott** [comment]
Yeah let's do it. Do you have time to tackle that? If not I might get around to it some point this weekend.

**Comment by oliverlee** [comment]
> Yeah let's do it. Do you have time to tackle that? If not I might get around to it some point this weekend.

Maybe this afternoon? If not, also this weekend.

**Comment by crtrott** [comment]
Sounds good. Also I send you an invite to our slack channel in case you are interested.

**Comment by crtrott** [comment]
Note: I merged one other PR, but this one rebases cleanly on top of it.

**Comment by oliverlee** [comment]
Rebased + squashed some commits + separated out the death tests. Is this what you had in mind?

**Comment by crtrott** [comment]
Yeah this is great. Thanks so much!

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/macros.hpp]: I think this needs to be decorated with `MDSPAN_FUNCTION`

Comment (dalg24) [reply include/experimental/__p0009_bits/macros.hpp]: Did you consider tag dispatching instead ,having two non templated overloads
```C++
MDSPAN_FUNCTION constexpr void precondition(std::false_type /*checked*/, const char* /*cond*/, const char* /*file*/, unsigned /*line*/) { /* do nothing */ }
MDSPAN_FUNCTION constexpr void precondition(std::true_type /*checked*/, const char* cond, const char* file, unsigned line) { MDSPAN_IMPL_PRECONDITION_VIOLATION_HANDLER(cond, file, line); }

File: include/experimental/__p0009_bits/utility.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/utility.hpp:59]: Where are these used?

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/macros.hpp]: Fine by me but just noting that the default handler would not work with CUDA/HIP

File: include/experimental/__p0009_bits/utility.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/utility.hpp:59]: line 526 in layout_stride.hpp and 546 or so in extents.hpp respectively.

File: include/experimental/__p2642_bits/layout_padded_fwd.hpp
Comment (dalg24) [inline include/experimental/__p2642_bits/layout_padded_fwd.hpp]: Should we not be using `MDSPAN_IMPL_PRECONDITION` here?

File: tests/CMakeLists.txt
Comment (dalg24) [inline tests/CMakeLists.txt:25]: I am not necessarily opposed to this but IMO it does not belong to this PR

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (dalg24) [inline tests/test_alternate_precondition_violation_handler.cpp:6]: I am confused.  What is this suppose to test?
~~For one it is throwing regardless of whether the condition evaluates to false or not.~~ **EDIT: I see now that the condition is meant for debugging purposes, not for evaluation.  Sorry about the noise**
Also the `MDSPAN_IMPL_` denotes that that the macro is "private" or "internal" which indicates it is not meant as a customization point for users.

File: tests/test_macros.cpp
Comment (dalg24) [inline tests/test_macros.cpp:1]: Do we care to test that precondition checking can indeed be disabled?

File: include/experimental/__p0009_bits/macros.hpp
Comment (oliverlee) [reply include/experimental/__p0009_bits/macros.hpp]: I might have a while back? Not sure.

But I tested it out just now, and this approach produces warnings with Clang and GCC:

Clang-18:
```
.../mdspan/include/mdspan/../experimental/__p0009_bits/macros.hpp:142:16: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
  142 | constexpr void precondition(std::true_type, const char* cond, const char* file, unsigned line)
```

GCC-11:
```
.../mdspan/include/mdspan/../experimental/__p0009_bits/macros.hpp:120:81: error: call to non-‘constexpr’ function ‘void Kokkos::detail::default_precondition_violation_handler(const char*, const char*, unsigned int)’
```

File: include/experimental/__p0009_bits/macros.hpp
Comment (oliverlee) [reply include/experimental/__p0009_bits/macros.hpp]: Please feel free to make any changes if CUDA/HIP compatibility is necessary.

File: include/experimental/__p2642_bits/layout_padded_fwd.hpp
Comment (oliverlee) [reply include/experimental/__p2642_bits/layout_padded_fwd.hpp]: Sure let's do that.

File: tests/CMakeLists.txt
Comment (oliverlee) [reply tests/CMakeLists.txt:25]: This change is related to C++14 support to some degree - otherwise C++17 language features are still usable with some compilers when building for C++14. But I also see that the impact of this change has wider consequences.

What do you think about breaking out this change into a separate PR and rebasing this PR on top of that? I can do that later this evening - or feel free to change this PR as you see fit.

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (oliverlee) [reply tests/test_alternate_precondition_violation_handler.cpp:6]: No worries about the noise. Should there be some documentation here for future readers?

File: tests/test_macros.cpp
Comment (oliverlee) [reply tests/test_macros.cpp:1]: Yes that's probably a good idea.

I'm not sure what's the best way - possibly a new test file+target that defines a precondition handler that unconditionally throws and then setting `MDSPAN_IMPL_CHECK_PRECONDITION 0`?

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [reply include/experimental/__p0009_bits/macros.hpp]: I expect the issue is the `constexpr` specifier

Comment (dalg24) [reply include/experimental/__p0009_bits/macros.hpp]: I suppose you needed it to be call it in other functions that themselves are callable in a constant expression...

File: tests/test_alternate_precondition_violation_handler.cpp
Comment (crtrott) [reply tests/test_alternate_precondition_violation_handler.cpp:6]: The original idea was to have the macros public. But I wanted to consider that a bit more in-depth and thus said to make it private for now. We should still test the mechanism.

Comment (dalg24) [reply tests/test_alternate_precondition_violation_handler.cpp:6]: Sure

File: include/experimental/__p0009_bits/macros.hpp
Comment (mhoemmen) [reply include/experimental/__p0009_bits/macros.hpp]: `printf` works just fine with CUDA; it's HIP that's broken lol ; - )

File: include/experimental/__p0009_bits/macros.hpp
Comment (mhoemmen) [inline include/experimental/__p0009_bits/macros.hpp]: I don't know how HIP or SYCL work, but the following should be fine for CUDA.  Note that in device code, `printf` may have a performance cost even if it is never called at run time.  (For example, the compiler may need to reserve registers for it.)

```suggestion
{
#if defined(__CUDA_ARCH__)
#  if ! defined(NDEBUG)
  printf("%s:%u: precondition failure: `%s`\n", file, line, cond);
  assert(false);
#  endif
#else
  fprintf(stderr, "%s:%u: precondition failure: `%s`\n", file, line, cond);
  assert(false);
#endif 
}
```

Comment (masterleinad) [reply include/experimental/__p0009_bits/macros.hpp]: `HIP` works the same as `Cuda`. For SYCL (oneAPI), you would need to use `sycl::ext::oneapi::experimental::printf`.

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [reply include/experimental/__p0009_bits/macros.hpp]: it is a `fprintf` and it is qualified so there is no way it works on CUDA.  As far as I know HIP behaves just the same as CUDA here.

File: tests/CMakeLists.txt
Comment (oliverlee) [reply tests/CMakeLists.txt:25]: https://github.com/kokkos/mdspan/pull/328

File: include/experimental/__p0009_bits/layout_left.hpp
Comment (nmm0) [inline include/experimental/__p0009_bits/layout_left.hpp:137]: This should just be guarded by C++17. We shouldn't include the layout_padded_fwd.hpp header in C++14 mode

File: include/experimental/__p2642_bits/layout_padded_fwd.hpp
Comment (nmm0) [inline include/experimental/__p2642_bits/layout_padded_fwd.hpp:86]: Let's just not include this header in C++14, since it's only used for padded layouts which require at least C++17

File: include/experimental/__p0009_bits/layout_right.hpp
Comment (nmm0) [inline include/experimental/__p0009_bits/layout_right.hpp:24]: Same here, let's guard this with `MDSPAN_HAS_CXX_17`

**Review by nmm0 — DISMISSED** [review]
I'm good with almost everything, just lets keep the padded layout stuff isolated (since it requires C++17 anyway)

**Comment by crtrott** [comment]
I just noticed that this requires CXX Extensions for GCC. I.e. -std=gnu++14 - it doesn't work with -std=c++14. Do you consider that acceptable?

File: include/experimental/__p0009_bits/layout_left.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_left.hpp:137]: done

**Comment by oliverlee** [comment]
> I just noticed that this requires CXX Extensions for GCC. I.e. -std=gnu++14 - it doesn't work with -std=c++14. Do you consider that acceptable?

Do you know what requires extensions? I didn't intend for that and I believe I am only using -std=c++14 in projects using mdspan (however, those don't use CMake).

**Comment by crtrott** [comment]
Its now fixed.

File: include/experimental/__p0009_bits/layout_stride.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/layout_stride.hpp:495]: No warning from `is_exhaustive_impl` not having the host device markups?

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/macros.hpp:25]: Please confirm that you mean to include this header and not `<cassert>`

File: include/experimental/__p0009_bits/macros.hpp
Comment (dalg24) [inline include/experimental/__p0009_bits/macros.hpp:116]: Please confirm that you did not see issue with Clang as the CUDA compiler.

File: tests/CMakeLists.txt
Comment (dalg24) [reply tests/CMakeLists.txt:25]: Christian did you mean to keep this?

File: tests/test_layout_preconditions.cpp
Comment (dalg24) [inline tests/test_layout_preconditions.cpp:27]: Wait why don't we special cases these?  Is the error message lost?

File: tests/test_mdspan_size.cpp
Comment (dalg24) [inline tests/test_mdspan_size.cpp:58]: Yuk

File: tests/test_macros.cpp
Comment (dalg24) [inline tests/test_macros.cpp:11]: I would have done
```c++
#if MDSPAN_IMPL_CHECK_PRECONDITION
ASSERT_DEATH(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"), msg);
#else
ASSERT_NO_THROW(MDSPAN_IMPL_PRECONDITION(false and "hello, world!"));
#endif
```
instead of messing with the test success criterion in the cmake file.

File: tests/test_layout_preconditions.cpp
Comment (crtrott) [reply tests/test_layout_preconditions.cpp:27]: Yeah because we don't use all the __assertfail crap.

File: tests/test_macros.cpp
Comment (dalg24) [reply tests/test_macros.cpp:11]: Please ignore.   This doesn't apply here.  I is only for `test_precondition_checks_can_be_disabled` Sorry about the noise.

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/macros.hpp:25]: I couldn't find a combo which fails. For Kokkos we anyway gonna replace the thing with Kokkos::abort.

File: include/experimental/__p0009_bits/layout_stride.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/layout_stride.hpp:495]: we are currently using constexpr-experimental stuff for mdspan, we will fix that later.

Comment (dalg24) [reply include/experimental/__p0009_bits/layout_stride.hpp:495]: fine

File: include/experimental/__p0009_bits/macros.hpp
Comment (crtrott) [reply include/experimental/__p0009_bits/macros.hpp:116]: Yeah I couldn't find a variant which fails and we will use in Kokkos Kokkos::abort anyway.

**Review by dalg24 — APPROVED** [review]
Good enough.
Thank you Oliver.

**Pull request merged**
